### PR TITLE
adding gatk-launch script to make it easier to run spark

### PR DESCRIPTION
**Pull request by lbergelson**
this is a script which can be used after running gradle installDist to run spark jobs
it can be used identically to ths build/install/bin/gatk script, but has extra features for dealing with spark

running a spark tool and supplying the option --sparkTarget with LOCAL, CLUSTER, or GCS has special behavior
LOCAL will run the tool in the in memory spark runner
CLUSTER along with an appropriate --sparkMaster will run on an accessible spark cluster using spark-submit
arguments to spark-submit may be specified before the arguments to GATK by separating them with a --
GCS will submit jobs to google dataproc using gcloud
common arguments for spark submit will be adapted to match the gcloud formating
this will fail if gcloud isn't installed

if GATK_GCS_STAGING is specified, the jar will be uploaded and cached in the specified bucket for rapid re-use

input files will not be autouploaded to the cloud

--dry-run may be specified before the --, this will only print the commands that will be run instead of actually running them

Adding DataProcArgumentReplace simple tool to convert spark-submit args into gcloud args.
This conversion is not guarenteed to translate all spark command line options to matching gcloud ones.
If you find options that are not translated or are miss-translated please file an issue.

### CODE REVIEW

**Comment by lbergelson** [comment]
Hmn. for some reason the pull is failing when the push succeeds...

**Comment by droazen** [comment]
Since this is going to be THE way to run the GATK for now (walkers, picard tools, and spark tools), I think we should rename and move your script. `gatk-submit.sh` sounds too spark-specific, and the `.sh` extension is unnecessary (users don't need to know they're running a shell script rather than a binary). I'd suggest something more generic like `gatk-launch` (no extension), and move the script to the root of our repo.

**Comment by tomwhite** [comment]
This needs a help option/usage information as it's not obvious how to construct the options...

File: build.gradle [hunk 1]
```diff
@@ -400,6 +396,17 @@ uploadArchives {
...
             throw new Exception("Missing sonatype username or password, please see the README for properties that must be specified for uploading an archive.")}
         )
-
     }
+
+distributions {
+    spark{} //added to allow an installSparkDist target which will also build the spark jar
+}
+
+installSparkDist.dependsOn downloadGsaLibFile, sparkJar
```
Comment (droazen) [inline build.gradle]: `installSparkDist`? This is not quite the easy-to-remember name I was hoping for for the "build everything" target. How about: `gradle installDist` behaves as it does currently, we rename `gradle sparkJar` to `gradle installSpark`, and we add a `gradle installAll` that does both? That way, we have `gradle installDist`, `gradle installSpark`, and `gradle installAll`, which is at least consistent naming-wise. Thoughts from others on the team?

Comment (tomwhite) [inline build.gradle]: Is there a Gradle convention here? In Maven it would be `install`.

Comment (lbergelson) [inline build.gradle]: So as far as I understand, the gradle local ablation default installDist, because the early maven plug in stole install.

Comment (droazen) [inline build.gradle]: Ablation, eh? Someone is on a phone :) Well, if we can't use `gradle install`, what do you guys think about my proposal of having `gradle installDist`, `gradle installSpark`, and `gradle installAll`?

**Comment by tomwhite** [comment]
Why did you switch from bash? Using Python adds another runtime dependency, which will likely increase the number of support tickets. Ideally the script would be fairly dumb and pass through its arguments, so there shouldn't be a lot of complexity for bash to handle.

I think that the settings in `SparkContextFactory#CMDLINE_ATTRIBUTES` should be removed since they can't be overridden as it stands. If they are specified in the launcher script only (as they are with this PR) then they can be overridden by adding a `--conf` argument to the Spark args - this works since the last conf value wins.

I would leave `spark.serializer` and `spark.kryo.registrator` in `SparkContextFactory#CMDLINE_ATTRIBUTES` since they are required for correctness, but remove all the others. For example, it's certainly conceivable that `spark.driver.userClassPathFirst`, which we've talked about before as always being required, may have to be overridden for certain versions of Spark.

**Comment by droazen** [comment]
@tomwhite @lbergelson tried writing this in bash, but it ended up being quite messy due to lack of things like maps, so we decided to switch to python. The script does need to do a few non-trivial things like translation between `spark-submit` and `gcloud` syntax for spark properties. I think a python dependency for running the front-end script is reasonable these days, as python is fairly ubiquitous -- particularly as nothing will force people to use this script (can always use the jar + `spark-submit` or `gcloud` directly if desired).

File: build.gradle (hunk 1 above)
Comment (droazen) [inline build.gradle]: Alternatively, we could make `gradle installDist` do everything (local dist + spark jar + other things as we add them), if that is more consistent with gradle convention, and have `gradle installSpark` for just the spark jar and `gradle installLocal` (or something like that) for just the unpackaged local version.

**Comment by tomwhite** [comment]
> nothing will force people to use this script

No, but people starting out will use it so it's important that it works.

Which versions of Python are supported?

**Comment by droazen** [comment]
@tomwhite I believe @lbergelson's script will work on both Python 2.x and 3.x, but I'll let him answer that definitively.

**Comment by lbergelson** [comment]
Should work on 2.7.x and 3.x now.

**Comment by lbergelson** [comment]
@tomwhite If we remove those hardcoded values then any tool that builds ontop of hellbender will need to know to add them themselves.  I know this was a major confusion for hellbender protected.

**Comment by droazen** [comment]
Review in person with @lbergelson -- :+1: merge when ready.

**Comment by lbergelson** [comment]
@laserson I've tested using python 2.6, 2.7 and 3.5.0  All seem to work.

**Comment by tomwhite** [comment]
I tested this successfully on a cluster. The changes make it possible to override Spark configuration - e.g. I was able to override `spark.yarn.executor.memoryOverhead`.

**Comment by akiezun** [comment]
@lbergelson is this ready to merge?

**Pull request merged**